src/validate.ts

```diff
@@ -26,18 +35,64 @@ function responseValidation(apiResponse: object, schema: object) {
 
 /**
  * Identifies the response type and validates the function based on the schemas.
- * @param apiResponse User defined api response
- * @param responseType User defined intent response
+ * @param req Request from one of the generate request functions.
+ * @param apiResponse User defined api response.
  * @return Errors from AJV validation, if any. Undefined otherwise.
  */
 export function validate(intentRequest: object, apiResponse: object) {
   const responseType = intentRequest['inputs'][0]['intent'];
   if (responseType === 'action.devices.SYNC') {
-    return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+    // validate traits with attributes schema
+    const syncErrors : object[] = [];
+    const syncDevices = apiResponse['payload']['devices'];
+    const syncDevicesLength = syncDevices.length;
+    for (let i = 0; i < syncDevicesLength; i++) {
+      const traits = syncDevices[i]['traits'];
+      const traitsLength = syncDevices[i]['traits'].length;
+      const attributes = syncDevices[i]['attributes'];
+      if (attributes === undefined) {
+        return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+      }
+      for (let j = 0; j < traitsLength; j++) {
+        const trait = traits[j];
+        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
+          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+          if (validateTraitRes) {
+            return syncErrors.push(...validateTraitRes);
+          }
+        }
+      }
+    }
+    return syncErrors.length ? syncErrors : undefined;
   } else if (responseType === 'action.devices.QUERY') {
+    // validate with states schema
     return responseValidation(apiResponse, QUERY_RESPONSE_SCHEMA);
   } else if (responseType === 'action.devices.EXECUTE') {
-    return responseValidation(apiResponse, EXECUTE_RESPONSE_SCHEMA);
+    // validate with states schema
+    const executeErrors : object[] = [];
+    // gets the execution array from the intent request
+    const execution = intentRequest['inputs'][0]['payload']['execution'];
+    const executionLength = intentRequest['inputs'][0]['payload']['execution'].length;
+
+    // identifies the part of the api response to validate against a schema
+    const commands = apiResponse['payload']['commands'];
+    const commandsLength = commands.length;
+
+    for (let i = 0; i < executionLength; i++) {
+      // gets the specific command
+      const commandName = execution[i]['command'];
+      if (commandName in COMMAND_STATES_EXPECT) {
```

Comment:
Nit: I would argue if this is false, it should be considered an error / invalid?